[docs] Update the migration guide with the breaking changes between the legacy and the new pickers (#7345)
mui/mui-x: docs/data/migration/migration-pickers-v5/migration-pickers-v5.md

All other changes must be handled manually.

## Picker components

> ### New picker components

Author: This section would be removed when removing the legacy pickers
It is mainly here to help people trying out the beta understand what is going on

});
```

> ### Remove the keyboard view

Author: If we want to name the sections with stuff people have to actually do, then I don't know how to name this one

The picker components no longer have a keyboard view to render the input inside the modal on mobile.

- If your date is easier to edit with the keyboard (e.g: a birthdate), you can directly use the new field components:

  ```diff
   function App() {
     return (
  -    <DatePicker />
  +    <DateField />
     )
   }
  ```

- If you want users to be able to edit with both the keyboard using the input and with the touch using the views,
  you can pass a custom `Layout` component slot to re-introduce the keyboard view.

  :::warning

>   TODO: Add a demo in the "Custom layout" page to add it back.

Author: Need #7347

  :::

  :::info
  At some point, the mobile pickers should have a prop allowing to have an editable field without opening the modal.
  :::

> ### Change the DOM structure

Author: For these changes, maybe a few example with the DOM structure before / after would help
Because we don't have a 1 to 1 equivalent of DOM nodes since we reduced the nesting and moved to a grid approach.

We could create an example to add back the old DOM structure but I doubt it's a good solution.

Reviewer: I also don't think that we should endorse keeping the previous/flawed DOM structure. 🤔 

But I'm leaning towards adding this "breaking change" into codemod umbrella issue and considering at least adding a tranformation doing what you've listed here. 🤔 
Would you agree that it would make sense?

Author: I don't know if it's worth it.
Because most of those changes are not purely 1 to 1 equivalent.
And I fear the codemod are going to be a massive pain building.

     components: {
  -    MuiCalendarOrClockPicker: {
  +    MuiPickersLayout: {
         styleOverrides: {
  -        root: {
  +        contentWrapper: {
             backgroundColor: 'red',
           },
         },
       },
     },
   });
  ```

  :::warning
  These changes only apply to the new pickers.
  The legacy pickers keep the `CalendarOrClockPicker` internal component until there removal.

  For more information about those new pickers, take a look at the [New picker components](#new-picker-components) section
  :::

- The internal `PickerStaticWrapper` component has been removed and all its element have been moved to the new `Layout` component slot

  ```diff
   const theme = createTheme({
     components: {
  -    MuiPickerStaticWrapper: {
  +    MuiPickersLayout: {
         styleOverrides: {
           root: {
             opacity: 0.5,
           },
         },
       },
     },
   });
  ```

  The DOM node containing the toolbar and the view content (the `content` slot of the `PickerStaticWrapper` component) no longer exists.
  The closest equivalent is now the `contentWrapper` slot of the `PickersLayout` component, which do not contain the toolbar.

>   If you need a DOM node containing the toolbar and the view content, you will have to pass a [custom `Layout` component slot](/x/react-date-pickers/custom-layout/#dom-customization)

Reviewer: ```suggestion
  If you need a DOM node containing the toolbar and the view content, you will have to pass a [custom `Layout` component slot](/x/react-date-pickers/custom-layout/#dom-customization).
```

  ```diff
   const theme = createTheme({
     components: {
  -    MuiPickerStaticWrapper: {
  +    MuiPickersLayout: {
         styleOverrides: {
  -        content: {
  +        contentWrapper: {
             opacity: 0.5,
           },
         },
       },
     },
   });
  ```

  :::warning
  These changes only apply to the new pickers.
  The legacy pickers keep the `PickerStaticWrapper` internal component until there removal.

>   For more information about those new pickers, take a look at the [New picker components](#new-picker-components) section

Reviewer: ```suggestion
  For more information about these new pickers, take a look at the [New picker components](#new-picker-components) section
```
Or `the` 🤔

> ## Component slots (custom sub-components)

Author: We should probably re-order the sections inside `Component slots`) 
Either alphabetically or thematically

 </LocalizationProvider
```

## Component slots / component props slots

All the props used to pass props to parts of the UI (e.g: pass a prop to the input) have been replaced by components props slots.
All the props used to override parts of the UI (e.g: pass a custom day renderer) have been replaced by components slots.

You can find more information about this pattern in the [MUI Base documentation](https://mui.com/base/getting-started/usage/#shared-props).

> There changes always apply to all the components that had the prop.

Reviewer: ```suggestion
These changes apply to all the components that had the prop.
```

For example, the `ToolbarComponent` has been replaced by a `Toolbar` component slot on all pickers.

> ### Input renderer (required in v5)

Reviewer: Do you think we should consider having a codemod taking care of at least the basic examples in this section?

Author: I don't know how much of a pain it would be honestly :grimacing:

Reviewer: Well, we can always experiment with it if we'll have the time for it.
Or, get back to it if we'd have a lot of community requests. 🤔

- The `renderInput` has been replaced by an `input` component props slot:

  ```diff
   <DatePicker
  -  renderInput={(inputProps) => <TextField {...props} variant="outlined" />}
  +  componentsProps={{ input: { variant: 'outlined' }}
   />

   <DateRangePicker
  -  renderInput={(startProps, endProps) => (
  -    <React.Fragment>
  -      <TextField {...startProps} variant="outlined" />
  -      <Box sx={{ mx: 2 }}> - </Box>
  -      <TextField {...endProps} variant="outlined" />
  -    </React.Fragment>
  -  )}
  +  componentsProps={{ input: { variant: 'outlined' }}
   />
  ```

>   The Date Range Picker also have a new `fieldSeparator` component slot and component props slot to customize only this part of the UI:

Reviewer: Either this should also be a sub-section (list item), or we need to drop it from the top paragraph.

```suggestion
- The Date Range Picker also have a new `fieldSeparator` component slot and component props slot to customize only this part of the UI:
```

  ```diff
   <DateRangePicker
  -  renderInput={(startProps, endProps) => (
  -    <React.Fragment>
  -      <TextField {...startProps} />
  -      <Box sx={{ mx: 2 }}> to </Box>
  -      <TextField {...endProps} />
  -    </React.Fragment>
  -  )}
  +  componentsProps={{ fieldSeparator: { children: 'to' }}
   />
  ```

  :::warning
  This change only applies to the new pickers.
  The legacy pickers keep the `renderInput` prop until there removal.

>   For more information about those new pickers, take a look at the [New picker components](#new-picker-components) section

Reviewer: ```suggestion
  For more information about these new pickers, take a look at the [New picker components](#new-picker-components) section
```

  :::

### Toolbar (`ToolbarComponent`)

- The `ToolbarComponent` has been replaced by a `Toolbar` component slot:

  ```diff
   <DatePicker

>   -  ToolbarComponent={MyToolbar}

Reviewer: 🤦‍♂️ 🤣

 </LocalizationProvider
```

> ## Component slots / component props slots

Reviewer: ```suggestion
## Component slots / component slots props
```
If we go with the recent discussion result. 😃

> All the props used to pass props to parts of the UI (e.g: pass a prop to the input) have been replaced by components props slots.

Reviewer: ```suggestion
All the props used to pass props to parts of the UI (e.g: pass a prop to the input) have been replaced by component slots props.
```

> All the props used to override parts of the UI (e.g: pass a custom day renderer) have been replaced by components slots.

Reviewer: ```suggestion
All the props used to override parts of the UI (e.g: pass a custom day renderer) have been replaced by component slots.
```

   />
  ```

### Input

- The `InputProps` prop has been removed.
  You can use the `InputProps` of the `input` component props slot instead:

  ```diff
   <DatePicker
  -  InputProps={{ color: 'primary' }}
  +  componentsProps={{ input: { InputProps: { color: 'primary' }}}}
   />
  ```

  :::warning
  This change only applies to the new pickers.

>   The legacy pickers keep the `InputProps` prop until there removal.
> 
>   For more information about those new pickers, take a look at the [New picker components](#new-picker-components) section

Reviewer: ```suggestion
  The legacy pickers keep the `InputProps` prop until their removal.

  For more information about these new pickers, take a look at the [New picker components](#new-picker-components) section.
```

  :::

### Input adornment

> - The `InputAdornmentProps` prop has been replaced by a `inputAdornment` component props slot:

Reviewer: General nitpick: The prop is named `componentsProps`, but in the most part of the documentation we are calling it `component props` instead of `components props`. Maybe it would be worth keeping those more in sync between the prop name and usage? 🤔

Author: So you would work it `components props slot`, make sense to me

Reviewer: Hm. Technically, I'd consider `slot` a place where I can put/replace something. 🤔 
In case of `components props` it's not really a slot.
What do you think about using `components slot(s)` and `components props` in the documentation as well?
Basically, drop `slot` where we are talking about props and not the component slot itself...

Author: `component slots props` ?
When we will rename the prop `slotsProps`, won't it be weird if we don't have the word `slot` at all in the sentence ?

Reviewer: Fair point. 🤔 
I thought that we could change these at the same time when we move towards `slots` and `slotsProps`, but on the other hand, just sticking to `component slots props` probably seems the best...
I do agree with your suggestion. 👍

Author: Applied

  ```diff
   <DatePicker
  -  InputAdornmentProps={{ position: 'start' }}
  +  componentsProps={{ inputAdornment: { position: 'start' }}}
   />
  ```

  :::warning
  This change only applies to the new pickers.

>   The legacy pickers keep the `InputAdornmentProps` prop until there removal.
> 
>   For more information about those new pickers, take a look at the [New picker components](#new-picker-components) section

Reviewer: ```suggestion
  The legacy pickers keep the `InputAdornmentProps` prop until their removal.

  For more information about these new pickers, take a look at the [New picker components](#new-picker-components) section.
```

  :::

### Open Picker Button

- The `OpenPickerButtonProps` prop has been replaced by a `openPickerButton` component props slot:

  ```diff
   <DatePicker
  -  OpenPickerButtonProps={{ ref: buttonRef }}
  +  componentsProps={{ openPickerButton: { ref: buttonRef }}}
   />
  ```

  :::warning
  This change only applies to the new pickers.

>   The legacy pickers keep the `OpenPickerButtonProps` prop until there removal.
> 
>   For more information about those new pickers, take a look at the [New picker components](#new-picker-components) section

Reviewer: ```suggestion
  The legacy pickers keep the `OpenPickerButtonProps` prop until their removal.

  For more information about these new pickers, take a look at the [New picker components](#new-picker-components) section.
```
